### PR TITLE
feat: Add Delegate type

### PR DESCRIPTION
**Pull request by paulgessinger**
This type is intended as a sort-of replacement of `std::function` without the virtual function overhead of that type. This is ultimately meant to support the new extension mechanism for the KF and CKF, where these delegates will be used for the calibrator, smoother, and so on.

Delegate type that allows type erasure of a callable without allocation and with a single level of indirection. This type can support:
- a free function pointer
-  - a pointer to a member function alongside an instance pointer
Note: `Delegate` does not assume ownership of the instance. You need to ensure that the lifetime of the callable   instance is longer than that of the @c Delegate.
Currently `Delegate` only supports callables that are `const`.

The usage of this type looks like this for a free function:

```cpp
int sumImplementation(int a, int b) {
  return a + b;
}

Delegate<int(int, int)> sum;
sum.connect<&sumImplementation>();

int c = sum(2, 2); // = 4
```

and like this for a member pointer with instance:

```cpp
struct Subtractor {
  int x;

  int subtract(int v) const {
    return v - x;
  }
};

Delegate<int(int)> subtract;

Subtractor instance{5};

subtract.connect<&Subtractor::subtract>(&instance);

int y = subtract(10); // = 5
```

### CODE REVIEW

File: Core/include/Acts/Utilities/Delegate.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  /// Connect a free function pointer.
+  /// @note The function pointer must be ``constexpr`` for @c Delegate to accept it
+  /// @tparam Callable The compile-time free function pointer
+  template <auto Callable>
+  void connect() {
+    m_payload = nullptr;
+    m_function = [](const void* /*payload*/, Args... args) -> return_type {
+      return std::invoke(Callable, std::forward<Args>(args)...);
+    };
```
Comment (stephenswat) [inline Core/include/Acts/Utilities/Delegate.hpp:48]: Out of curiosity, doesn't this defeat the purpose of reducing the number of indirections? I imagine that if you call the delegate object, you're retrieving the function pointer to this lambda first, then the lambda invokes the callable, isn't that two levels of indirections? Or am I misunderstanding your design?

Comment (paulgessinger) [reply Core/include/Acts/Utilities/Delegate.hpp:48]: The other function pointer is a compile time function pointer. You're right that it's not **guaranteed** to be reduced to a single one, but I would think in almost all cases it should be.

Comment (paulgessinger) [reply Core/include/Acts/Utilities/Delegate.hpp:48]: I could probably double check this in godbolt though.

Comment (paulgessinger) [reply Core/include/Acts/Utilities/Delegate.hpp:48]: So my [godbolt check](https://godbolt.org/z/zxb5653z8) indicates that giving it the function pointer at compile-time should almost make sure the extra lambda is removed;

**Comment by timadye** [comment]
I've also been playing with this in [Godbolt](https://godbolt.org/z/aGsc5xE1b). First a couple of minor tweaks:
 1. `Delegate.hpp` needs `#include <cassert>`
 2. in your example above (and your slides) need to declare `Subtractor::subtract() const` (took me a long time to work that one out!).

I compared `Delegate` with a couple of implementations using `std::function`.

  1. With GCC 11, `Delegate` generates obviously much more efficient code than `std::function`. If the implementation is known (uncomment definitions on Godbolt) the `Delegate` example can be fully resolved to a literal number (9)!
 2. Clang 13 is even cleverer and can inline everything. the `Delegate` and `std::function` give identical object code, whether the function definitions are known or not.

Anyway, since GCC isn't yet clever enough to optimise the `std::function` call away, `Delegate` looks like a useful optimisation.

To my eye, `Delegate::connect` (eg. `s.connect<&Subtractor::subtract>(&inst)`) seems a little counterintuitive. Is there an (equally efficient and compact) way to provide a constructor and assignment operator to do the same thing. I guess the problem is passing the member function address as a type, but maybe there's a way.

Thanks and sorry for the random comments,
Tim.

**Comment by paulgessinger** [comment]
Hey @timadye, no worries I'm super happy about comments like this! I hadn't looked at the objects generated via `std::function`, to be honest, only at descriptions of how it's implemented. I guess in the long run, this wouldn't be needed then. You're right that I forgot about the `const` qualifiers.

In terms of the interface: yes, I'm not sure this works with an assignment operator, since the callable has to be given as a template parameter. Lambas with captures cannot be converted to a function pointer, so the only reason I'm able to "capture" the target callable is because it's a compile time parameters. So even an assignment operator which gives the function pointer at runtime wouldn't work, since I can't generate the wrapping lambda that discards the empty instance pointer.

Now, I have one overload 

```cpp
void connect(function_type callable);
```

already added which takes a callable at runtime, but that callable explicitly has to accept an `const void*` as the first argument. So this pattern is already not necessarily great, but if you feel like it improves clarity, I can replace/supplement this with an assignment operator that accepts the same callable type.

**Comment by timadye** [comment]
Hi @paulgessinger,

Thanks for the interesting discussion. I think the `connect` method is OK, just thought maybe it was worth exploring perhaps-clearer interfaces - especially if `Delegate` becomes a widely used type. We can stop if this is delaying more important developments.

How about this idea: implement a `make_delegate` function (or better name?) and assignment operator for the result, so you could say something like:

```
subtract = make_delegate<&Subtractor::subtract>(&instance);
```

and maybe also the copy constructor to initialise at construction:

```
Delegate<int(int)> subtract = make_delegate<&Subtractor::subtract>(&instance);
```

That seems clearer to me, but maybe it's just needless tinkering?

I don't know whether `make_delegate` could return a temporary `Delegate` object (inferring the `<int(int)>`, or would have to return a transfer type. In either case the temporary should be optimised away.

Even nicer would be to have a `Delegate<int(int),&Subtractor::subtract>()` constructor, but I guess that's not possible. Also I suppose if we wanted to use type deduction and pass `&Subtractor::subtract` as an argument, we'd defeat the whole purpose of your optimisation.

What do you think?
Thanks,
Tim.

**Comment by paulgessinger** [comment]
Hey @timadye, I've added a constructor and an assignment operator from runtime function pointer. I could add a `make_delegate` function and operators, but that to me just feels like what I already have with `connect()` but with different symbols. Do you feel strongly about this?

**Comment by timadye** [comment]
Hi @paulgessinger,

I'm not really sure how the `function_type` constructor etc will be used apart from to duplcate the lambda function call done by `connect()`. Is there a use case - apart from to implement my `make_delegate` suggestion?

I suggested `make_delegate` (name unimportant, but following an obvious pattern) simply to allow construction to look like construction and assignment to look like assignment - see my two examples [above](https://github.com/acts-project/acts/pull/1059#issuecomment-959169306). To me, `connect` isn't a pattern I'm used to, but maybe it's more familar to others.

I don't feel very strongly about any of this, just suggesting to explore some alternatives in case you found them useful. So if you prefer `connect`, go with that. 😄

Thanks,
Tim.

**Comment by timadye** [comment]
PS. were you planning to add the missing `#include <cassert>` ?

**Comment by paulgessinger** [comment]
I feel like having the extra constructor and operator is a good enough tradeoff, so I think I'd stick to this version.
If you're happy, can you approve?

**Review by timadye — APPROVED** [review]
sorry, I didn't know if I had permission to approve anything. Let's see if this works :)

**Review by timadye — APPROVED** [review]

**Pull request merged**